mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,772 @@
+from mava.wrappers import episode_metrics
+
+Metrics = Dict[str, Array]
+
+
+class Transition(NamedTuple):
+    obs: Array
+    action: Array
+    reward: Array
+    done: Array
+
+
+BufferState: TypeAlias = TrajectoryBufferState[Transition]
+
+
+class IQLParams(NamedTuple):
+    online: FrozenVariableDict
+    target: FrozenVariableDict
+
+
+class LearnerState(NamedTuple):
+    # Interaction vars
+    obs: Array
+    done: Array
+    hidden_state: Array
+    env_state: State
+    time_steps: Array
+
+    # Train vars
+    train_steps: Array
+    opt_state: optax.OptState
+
+    # Shared vars
+    buffer_state: TrajectoryBufferState
+    params: IQLParams
+    key: PRNGKey
+
+
+class ActionSelectionState(NamedTuple):
+    online_params: FrozenVariableDict
+    hidden_state: chex.Array
+    time_steps: int
+    key: chex.PRNGKey
+
+
+class InteractionState(NamedTuple):  # 'carry' in interaction loop
+    action_selection_state: ActionSelectionState
+    env_state: State
+    buffer_state: BufferState
+    obs: Observation
+    done: Array
+
+
+class TrainState(NamedTuple):  # 'carry' in training loop
+    buffer_state: BufferState
+    params: IQLParams
+    opt_state: optax.OptState
+    train_steps: Array
+    key: chex.PRNGKey
+
+
+class RecQNetwork(nn.Module):
+    num_actions: int
+    hidden_dim: int
+
+    def setup(self) -> None:
+        self.pre_torso: MLPTorso = MLPTorso((self.hidden_dim,))
+        self.post_torso: MLPTorso = MLPTorso((self.hidden_dim,))
+
+    @nn.compact
+    def __call__(
+        self,
+        hidden_state: chex.Array,
+        observations_resets: RNNObservation,
+    ) -> Array:
+        # unpack consumed parameters
+        obs, resets = observations_resets
+
+        embedding = self.pre_torso(obs.agents_view)
+
+        # pack consumed parameters
+        rnn_input = (embedding, resets)
+        hidden_state, embedding = ScannedRNN()(hidden_state, rnn_input)
+
+        embedding = self.post_torso(embedding)
+
+        q_values = nn.Dense(self.num_actions, kernel_init=orthogonal(0.01))(embedding)
+
+        return hidden_state, q_values
+
+
+def init(
+    cfg: DictConfig,
+) -> Tuple[
+    Tuple[Environment, Environment],
+    LearnerState,
+    RecQNetwork,
+    optax.GradientTransformation,
+    TrajectoryBuffer,
+    MavaLogger,
+    chex.PRNGKey,
+]:
+
+    logger = MavaLogger(cfg)
+
+    # init key, get devices available
+    key = jax.random.PRNGKey(cfg.system.seed)
+    devices = jax.devices()
+    n_devices = len(devices)
+
+    def replicate(x: Any) -> Any:
+        """First replicate the update batch dim then put on devices."""
+        x = jax.tree_map(lambda y: jnp.broadcast_to(y, (cfg.system.update_batch_size, *y.shape)), x)
+        return jax.device_put_replicated(x, devices)
+
+    # make envs
+    env, eval_env = environments.make(cfg)
+
+    # actions, agents, keysplits
+    num_actions = int(env.action_spec().num_values[0])
+    num_agents = env.action_spec().shape[0]
```

Comment:
Now we can access these directly:
```suggestion
   num_actions = env.actions_dim
    num_agents = env.num_agents
```